cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/query/FlowQueries.kt

```diff
@@ -0,0 +1,528 @@
+ *
+ */
+package de.fraunhofer.aisec.cpg.query
+
+import de.fraunhofer.aisec.cpg.graph.*
+import de.fraunhofer.aisec.cpg.graph.AccessValues
+import de.fraunhofer.aisec.cpg.graph.Node
+import de.fraunhofer.aisec.cpg.graph.statements.expressions.Literal
+import de.fraunhofer.aisec.cpg.graph.statements.expressions.Reference
+
+/**
+ * Determines how far we want to follow edges within the analysis. [maxSteps] defines the total
+ * number of steps we will follow in the graph (unlimited depth if `null`).
+ */
+sealed class AnalysisScope(val maxSteps: Int? = null)
+
+/**
+ * Only intraprocedural analysis. [maxSteps] defines the total number of steps we will follow in the
+ * graph (unlimited depth if `null`).
+ */
+class INTRAPROCEDURAL(maxSteps: Int? = null) : AnalysisScope(maxSteps)
+
+/**
+ * Enable interprocedural analysis. [maxCallDepth] defines how many function calls we follow at most
+ * (unlimited depth if `null`). [maxSteps] defines the total number of steps we will follow in the
+ * graph (unlimited depth if `null`).
+ */
+class INTERPROCEDURAL(val maxCallDepth: Int? = null, maxSteps: Int? = null) :
+    AnalysisScope(maxSteps)
+
+/** Determines in which direction we follow the edges. */
+enum class AnalysisDirection {
+    /** Follow the order of the EOG */
+    FORWARD,
+    /** Against the order of the EOG */
+    BACKWARD,
+    /** In and against the order of the EOG */
+    BIDIRECTIONAL,
+}
+
+/** Determines if the predicate must or may hold */
+enum class AnalysisType {
+    /**
+     * The predicate must hold, i.e., all paths fulfill the property/requirement. No path violates
+     * the property/requirement.
+     */
+    MUST,
+    /**
+     * The predicate may hold, i.e., there is at least one path which fulfills the
+     * property/requirement.
+     */
+    MAY,
+}
+
+/** Configures the sensitivity of the analysis. */
+enum class AnalysisSensitivity {
+    /** Consider the calling context when following paths (e.g. based on a call stack). */
+    CONTEXT_SENSITIVE,
+
+    /**
+     * Differentiate between fields, attributes, known keys or known indices of objects. This does
+     * not include computing possible indices or keys if they are not given as a literal.
+     */
+    FIELD_SENSITIVE,
+
+    /**
+     * Also consider implicit flows during the dataflow analysis. E.g. if a condition depends on the
+     * value we're interested in, different behaviors in the branches can leak data and thus, the
+     * dependencies of this should also be flagged.
+     */
+    IMPLICIT;
+
+    operator fun plus(other: AnalysisSensitivity) = arrayOf(this, other)
+
+    operator fun List<AnalysisSensitivity>.plus(
+        other: AnalysisSensitivity
+    ): Array<AnalysisSensitivity> {
+        return arrayOf(*this.toTypedArray(), other)
+    }
+
+    operator fun plus(other: Array<AnalysisSensitivity>): Array<AnalysisSensitivity> {
+        return arrayOf(*other, this)
+    }
+}
+
+fun dataFlowBase(
```

Comment:
The naming "base" is misleading here. Isn't this the "extended" version of the function and the other one is the simpler one. Although I don't know if that is really necessary when we can just have default parameters 🤷

Reply:
No, it's not necessary and the old functions are now deleted.